controllers/common.go

```diff
@@ -0,0 +1,117 @@
+package controllers
+
+import (
+	"encoding/json"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/aws/aws-sdk-go/service/ecr/ecriface"
+	"github.com/quintilesims/d.ims.io/models"
+	"github.com/zpatrick/fireball"
+)
+
+func listRepositories(e ecriface.ECRAPI) ([]string, error) {
+	repositories := []string{}
+
+	fn := func(output *ecr.DescribeRepositoriesOutput, lastPage bool) bool {
+		for _, repository := range output.Repositories {
+			repositories = append(repositories, aws.StringValue(repository.RepositoryName))
+		}
+
+		return !lastPage
+	}
+
+	if err := e.DescribeRepositoriesPages(&ecr.DescribeRepositoriesInput{}, fn); err != nil {
+		return nil, err
+	}
+
+	return repositories, nil
+}
+
+func getRepositoryPolicy(e ecriface.ECRAPI, repositoryName string) (*models.PolicyDocument, error) {
+	policyDoc := &models.PolicyDocument{}
+
+	getPolicyInput := &ecr.GetRepositoryPolicyInput{}
+	getPolicyInput.SetRepositoryName(repositoryName)
+	if err := getPolicyInput.Validate(); err != nil {
+		return nil, err
+	}
+
+	getPolicyOutput, err := e.GetRepositoryPolicy(getPolicyInput)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() != ecr.ErrCodeRepositoryPolicyNotFoundException {
+				return nil, err
+			}
+		} else {
+			return nil, err
+		}
+	}
+
+	if getPolicyOutput.PolicyText != nil && aws.StringValue(getPolicyOutput.PolicyText) != "" {
+		if err := json.Unmarshal([]byte(aws.StringValue(getPolicyOutput.PolicyText)), policyDoc); err != nil {
+			return nil, err
+		}
+	}
+
+	return policyDoc, nil
+}
+
+func setRepositoryPolicy(e ecriface.ECRAPI, repositoryName, policyText string) error {
```

Comment:
From above - Seems like just re-purposing this function a bit would allow us to remove most of the other helpers here. Looks like the callers only really ever need `set` since they grab all of the accounts prior to this call.

Reply:
agreed.

Reply:
I had a look at maximizing re-use but I'm not seeing a clear path to expand this function to re-purpose it. The problem is that there are 3 cases that require slightly different logic before setting the policy which seem better encapsulated in their own function:
- add all accounts to a single new repo's policy
- add single account to all repos' policies
- remove single account from all repos' policies
✅